src/Stratis.Bitcoin.IntegrationTests/Wallet/ColdWalletTests.cs

```diff
@@ -155,7 +155,7 @@ public void WalletCanMineWithColdWalletCoins()
                 Transaction transaction1 = stratisSender.FullNode.WalletTransactionHandler().BuildTransaction(CreateContext(stratisSender.FullNode.Network, new WalletAccountReference(WalletName, Account), Password, sendto.ScriptPubKey, amountToSend, FeeType.Medium, confirmations));
 
                 // Broadcast to the other node
-                stratisSender.FullNode.NodeService<WalletController>().SendTransaction(new SendTransactionRequest(transaction1.ToHex()));
+                stratisSender.FullNode.NodeController<WalletController>().SendTransaction(new SendTransactionRequest(transaction1.ToHex()));
```

Comment:
I think I understand what you are trying to do here.
Controller should practically only get registered if there is an api server (or RPC server running)

But we have a hack where we call controllers in tests. Maybe instead get access to the api self host DI to ask for controllers (similarly to how we store a the fullnode services provider).

Reply:
@dangershony The idea is to stop treating controllers as though they are singletons. The reason is that `ModelState` has a 1-to-1 relationship with the controller class.

The only time we want to explicitly create controller instances is during tests. All other times it's up to the api server to discover controllers (which we exposed by registering the application parts) and create individual temporary instances (possibly for multiple parallel clients) for serving requests.

Reply:
>The idea is to stop treating controllers as though they are services or singletons  

In this case just register them as instance types instead of as singletons.

But I agree on the other parts, they should be discoverable only by the api feature or RPC) in fact they are the situations where we registre them as singletons is a hack it was not like that.


>he only time we want to explicitly create controller instances is during tests.
In this case the above code should not be in the fullnode but as an extension on the common test helper.
`NodeController<WalletController>` should not be in the production code.

Reply:
Moved to `Stratis.Bitcoin.IntegrationTests.Common.FullNodeExtensions.cs`.

Reply:
see my latests comments, it should be enough to just remove addsingleton calls and even test should work as before

Reply:
yes but this change will still be needed even with your proposed change as controllers will no longer be registered anyway.